### DISCUSSION TITLE
Trace Generation with a Goal - Changing the Goal

### DISCUSSION THREAD

**Discussion by beckydvn** [AI-Planning/macq#120] [General]
Hi there,

So I'm trying to get the functionality for traces with dynamic goals working, but I'm getting some issues on what I believe to be the tarski side of things. I set up a simple test where I just take part of the original goal, just to see if I could successfully re-write the problem file and generate a plan. (I'm just testing with the blocks problem/domain right now): 
https://github.com/QuMuLab/macq/blob/3efb33ad767f9a62649445e2a23341c63652a766/macq/generate/pddl/generator.py#L220

When I try to compute a plan though, I get an error at this line, that `resp` does not generate a plan: https://github.com/QuMuLab/macq/blob/3efb33ad767f9a62649445e2a23341c63652a766/macq/generate/pddl/generator.py#L241
The issue being that there was an error parsing the problem because the domain types and problem types don't like up - the domain (original file) is using "default object" while the problem is using "object." I figure that tarski deals with default objects differently, which is causing the mismatch. However, when I tried to also regenerate the domain file in order to match with the format of the problem one, I then get the issue that the domain doesn't have any types at all (think it's caused by this?) 
https://github.com/aig-upf/tarski/blob/80a33dc5fe6ad2602b40b1c69f758b91188dd8e1/src/tarski/io/fstrips.py#L238
while problem is still using "object," so I get the same parsing error.

TLDR: main issue looks to be a typing inconsistency with tarski... ;P

**Comment by haz** [comment]
Can you post the both the working and broken PDDL files?

**Reply by beckydvn** [reply]
Hmm, doesn't seem to let me post PDDL files here.. I could try on Teams? Or commit the broken files?

**Reply by haz** [reply]
Triple ` for the code pasting section, then just copy/paste the raw PDDL.

**Reply by beckydvn** [reply]
```

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Domain file automatically generated by the Tarski FSTRIPS writer
;;; 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(define (domain BLOCKS)
    (:requirements :equality)
    (:types
        
    )

    (:constants
        
    )

    (:predicates
        (on ?x1 - object ?x2 - object)
        (ontable ?x1 - object)
        (clear ?x1 - object)
        (handempty )
        (holding ?x1 - object)
    )

    (:functions
        
    )

    

    
    (:action pick-up
     :parameters (?x - object)
     :precondition (and (clear ?x) (ontable ?x) (handempty ))
     :effect (and
        (not (ontable ?x))
        (not (clear ?x))
        (not (handempty ))
        (holding ?x))
    )


    (:action put-down
     :parameters (?x - object)
     :precondition (holding ?x)
     :effect (and
        (not (holding ?x))
        (clear ?x)
        (handempty )
        (ontable ?x))
    )


    (:action stack
     :parameters (?x - object ?y - object)
     :precondition (and (holding ?x) (clear ?y))
     :effect (and
        (not (holding ?x))
        (not (clear ?y))
        (clear ?x)
        (handempty )
        (on ?x ?y))
    )


    (:action unstack
     :parameters (?x - object ?y - object)
     :precondition (and (on ?x ?y) (clear ?x) (handempty ))
     :effect (and
        (holding ?x)
        (clear ?y)
        (not (clear ?x))
        (not (handempty ))
        (not (on ?x ?y)))
    )

)
```

**Reply by beckydvn** [reply]
```

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Instance file automatically generated by the Tarski FSTRIPS writer
;;; 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(define (problem BLOCKS-10-0)
    (:domain BLOCKS)

    (:objects
        a b c d e f g h i j - object
    )

    (:init
        (clear c)
        (clear f)
        (ontable f)
        (ontable i)
        (on d i)
        (on c e)
        (on g h)
        (on h a)
        (on a d)
        (on j b)
        (on e j)
        (on b g)
        (handempty )
    )

    (:goal
        (and (on j e) (and (on e h) (and (on h b) (and (on b a) (and (on a g) (on g i))))))
    )

    
    
    
)

```

**Reply by beckydvn** [reply]
^Tarski problem/domain

**Reply by beckydvn** [reply]
Original blocks domain:
```
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; 4 Op-blocks world
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(define (domain BLOCKS)
  (:requirements :strips)
  (:predicates (on ?x ?y)
	       (ontable ?x)
	       (clear ?x)
	       (handempty)
	       (holding ?x)
	       )

  (:action pick-up
	     :parameters (?x)
	     :precondition (and (clear ?x) (ontable ?x) (handempty))
	     :effect
	     (and (not (ontable ?x))
		   (not (clear ?x))
		   (not (handempty))
		   (holding ?x)))

  (:action put-down
	     :parameters (?x)
	     :precondition (holding ?x)
	     :effect
	     (and (not (holding ?x))
		   (clear ?x)
		   (handempty)
		   (ontable ?x)))
  (:action stack
	     :parameters (?x ?y)
	     :precondition (and (holding ?x) (clear ?y))
	     :effect
	     (and (not (holding ?x))
		   (not (clear ?y))
		   (clear ?x)
		   (handempty)
		   (on ?x ?y)))
  (:action unstack
	     :parameters (?x ?y)
	     :precondition (and (on ?x ?y) (clear ?x) (handempty))
	     :effect
	     (and (holding ?x)
		   (clear ?y)
		   (not (clear ?x))
		   (not (handempty))
		   (not (on ?x ?y)))))
```
and problem:
```
(define (problem BLOCKS-10-0)
(:domain BLOCKS)
(:objects D A H G B J E I F C )
(:INIT (CLEAR C) (CLEAR F) (ONTABLE I) (ONTABLE F) (ON C E) (ON E J) (ON J B)
 (ON B G) (ON G H) (ON H A) (ON A D) (ON D I) (HANDEMPTY))
(:goal (AND (ON D C) (ON C F) (ON F J) (ON J E) (ON E H) (ON H B) (ON B A)
            (ON A G) (ON G I)))
)
```

**Reply by haz** [reply]
Got it.

https://github.com/aig-upf/tarski/issues/113

**Reply by beckydvn** [reply]
Awesome!

**Reply by haz** [reply]
For now, can you manually add a type `object` and see if it dumps/solves ok?

**Reply by beckydvn** [reply]
I'll give it a shot... back to tarski finagling then ;)

**Reply by beckydvn** [reply]
Looks like I get an error when I try to change `lang.sorts` explicitly telling me I can't change the attribute 😆 So no luck..

**Reply by haz** [reply]
For now, rewrite the domain file slightly: replace `:fluents` with `:fluents object`. Silly, but it'll do the trick until I can tackle https://github.com/aig-upf/tarski/issues/113

**Reply by beckydvn** [reply]
So, I added the `object` type as you suggested (as well as the `typing` requirement because I think that's needed?) to the domain file. From what I can tell, the generated PDDL files look OK now. Here's the domain:

```
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Domain file automatically generated by the Tarski FSTRIPS writer
;;; 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(define (domain BLOCKS)
    (:requirements :equality :typing)
    (:types
		object - object        
    )

    (:constants
        
    )

    (:predicates
        (on ?x1 - object ?x2 - object)
        (ontable ?x1 - object)
        (clear ?x1 - object)
        (handempty )
        (holding ?x1 - object)
    )

    (:functions
        
    )
    
    (:action pick-up
     :parameters (?x - object)
     :precondition (and (clear ?x) (ontable ?x) (handempty ))
     :effect (and
        (not (ontable ?x))
        (not (clear ?x))
        (not (handempty ))
        (holding ?x))
    )


    (:action put-down
     :parameters (?x - object)
     :precondition (holding ?x)
     :effect (and
        (not (holding ?x))
        (clear ?x)
        (handempty )
        (ontable ?x))
    )

    (:action stack
     :parameters (?x - object ?y - object)
     :precondition (and (holding ?x) (clear ?y))
     :effect (and
        (not (holding ?x))
        (not (clear ?y))
        (clear ?x)
        (handempty )
        (on ?x ?y))
    )

    (:action unstack
     :parameters (?x - object ?y - object)
     :precondition (and (on ?x ?y) (clear ?x) (handempty ))
     :effect (and
        (holding ?x)
        (clear ?y)
        (not (clear ?x))
        (not (handempty ))
        (not (on ?x ?y)))
    )

)
```

And the problem:

```
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Instance file automatically generated by the Tarski FSTRIPS writer
;;; 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(define (problem BLOCKS-10-0)
    (:domain BLOCKS)

    (:objects
        a b c d e f g h i j - object
    )

    (:init
        (clear f)
        (clear c)
        (ontable i)
        (ontable f)
        (on b g)
        (on g h)
        (on d i)
        (on c e)
        (on h a)
        (on a d)
        (on j b)
        (on e j)
        (handempty )
    )

    (:goal
        (and (on d c) (on c f) (on f j) (on j e) (on e h) (on h b) (on b a) (on a g) (on g i))
    )
)
```

I haven't changed the goal at all here because I just want to get the planner to accept tarski-generated files first. Unfortunately it's still spitting out an error instead of a plan. Here's what I'm getting for `resp` here:
https://github.com/QuMuLab/macq/blob/193b26a481f2498f1e0c99d84c14fcf1c2f66305/macq/generate/pddl/generator.py#L274

`{'status': 'error', 'result': {'killed': False, 'code': 124, 'signal': None, 'cmd': 'timeout 5 python /app/process_solution.py /tmp/solver_planning_domains_tmp_4OrLguPQIZhnA/domain.pddl /tmp/solver_planning_domains_tmp_4OrLguPQIZhnA/problem.pddl /tmp/solver_planning_domains_tmp_4OrLguPQIZhnA/plan /tmp/solver_planning_domains_tmp_4OrLguPQIZhnA/log'}}`

Any clue as to what could be causing that?

**Reply by haz** [reply]
Hrmz...did you post-process the generated domain file, or add the `object` type via tarski? I'd suggest doing the former. So something like...

```python
with open('domain.pddl', 'r') as f:
    domain = f.read()

domain.replace(':types', ':types object')

with open('domain.pddl', 'w') as f:
    f.write(domain)
```
